### ISSUE TITLE
Unconditional failing when Floats equality is asserted breaks Elm Verify Examples

### ISSUE THREAD

**Issue by lzrski** [elm-explorations/test#88]
https://github.com/elm-explorations/test/blob/f42d322fdb524289b7edf91c4e68028389d862bb/src/Expect.elm#L747-L748

This creates some problems with `elm-verify-examples`. See https://github.com/stoeffel/elm-verify-examples/issues/83. Perhaps there could be a command line switch or something like `Expect.strictlyEqual` to allow to do this kind of comparison.

For reference here is a discussion that introduced this feature: https://github.com/elm-community/elm-test/issues/230

**Comment by mgold** [comment]
Oof. No obvious solution here so I'll invite brainstorming.

**Comment by drathier** [comment]
Floating point functions change depending on os, browser and cpu generation, so I don't think float equality is a good idea here either. Sooner or later someone will test a non-trivial float function, and it'll break. Maybe Stoeffel can add a second syntax for `Expect.within` checks, or leave out the ability to check floats?

**Comment by harrysarson** [comment]
There is another related issue where records that contain floating point values are really hard to test as you have to compare all keys one by one.

---

> Floating point functions change depending on os, browser and cpu generation

As an aside - I believe the IEEE floating point spec has in fact locked down the behaviour of floating point functions.

**Comment by drathier** [comment]
Yes, but IEEE754 is so relaxed that it allows for all these implementation-specific differences. I ran into this issue with elm-test in 2018, where I had unit tests for exact equality of floats, and the tests passed on mac but not on windows.

**Comment by lzrski** [comment]
> Maybe Stoeffel can add a second syntax for Expect.within checks, or leave out the ability to check floats?

Hmm... maybe something like:

```
0.1 --> 0.1 (+/- 0.0000001)
```

There was a related discussion here: https://github.com/stoeffel/elm-verify-examples/issues/39. Personally I agree with @stoeffel - these are just examples, so the additional syntax should be as straight forward as possible.

Perhaps it could be possible to detect floats (like this package is doing) and generate different test code using `within` instead of `equal`. Is there a tolerance that would always be safe? Meaning that it would allow for rounding errors but other than that would require equality? I don't think so, because the errors can accumulate, right?

**Comment by drathier** [comment]
Unfortunately, the errors build up. You'd have to specify both a relative (multiplicative) tolerance, and an absolute (additive) tolerance.

**Comment by mgold** [comment]
> Sooner or later someone will test a non-trivial float function, and it'll break.

Sure, but not as an example in the docs, right? If we can come up with a reasonable tolerance I think that would be ideal. Maybe allowing it to pass if within an absolute *or* a relative tolerance?

**Comment by drathier** [comment]
We can give it a try and see what happens ¯\_(ツ)_/¯ I think exact equality would be best in that case.

**Comment by lzrski** [comment]
> Sure, but not as an example in the docs, right? 

Why not? I mean the examples should be simple, but the internals of a tested function can be complex. I'm not saying it's impossible to come up with a reasonable tolerance, but the first assumption doesn't seem sound.

**Comment by pzp1997** [comment]
Perhaps we can use `elm-verify-examples.json` to specify the desired tolerance without altering the example syntax.
